File: kotlin/bc-envelope/src/main/kotlin/com/blockchaincommons/bcenvelope/FormatContext.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bctags.TAG_KNOWN_VALUE
import com.blockchaincommons.bctags.TAG_FUNCTION
import com.blockchaincommons.bctags.TAG_PARAMETER
import com.blockchaincommons.bctags.TAG_REQUEST
import com.blockchaincommons.bctags.TAG_RESPONSE
import com.blockchaincommons.bctags.TAG_EVENT
import com.blockchaincommons.dcbor.*
import com.blockchaincommons.knownvalues.KNOWN_VALUES
import com.blockchaincommons.knownvalues.KnownValue
import com.blockchaincommons.knownvalues.KnownValuesStore

/**
 * Option type for selecting which format context to use.
 */
sealed class FormatContextOpt {
    /** No format context. */
    data object None : FormatContextOpt()

    /** Use the global format context. */
    data object Global : FormatContextOpt()

    /** Use a custom format context. */
    data class Custom(val context: FormatContext) : FormatContextOpt()
}

/**
 * Context for formatting Gordian Envelopes with annotations.
 *
 * Provides information about CBOR tags, known values, functions, and parameters
 * that is used to annotate the output of envelope formatting functions.
 */
class FormatContext(
    private val tags: TagsStore = TagsStore(),
    private val knownValues: KnownValuesStore = KnownValuesStore(),
    private val functions: FunctionsStore = FunctionsStore(),
    private val parameters: ParametersStore = ParametersStore(),
) : TagsStoreTrait {

    /** Returns the CBOR tags registry. */
    fun tags(): TagsStore = tags

    /** Returns the known values registry. */
    fun knownValues(): KnownValuesStore = knownValues

    /** Returns the functions registry. */
    fun functions(): FunctionsStore = functions

    /** Returns the parameters registry. */
    fun parameters(): ParametersStore = parameters

    // -- TagsStoreTrait delegation --

    override fun assignedNameForTag(tag: Tag): String? = tags.assignedNameForTag(tag)

    override fun nameForTag(tag: Tag): String = tags.nameForTag(tag)

    override fun tagForName(name: String): Tag? = tags.tagForName(name)

    override fun tagForValue(value: ULong): Tag? = tags.tagForValue(value)

    override fun summarizer(tag: ULong): CborSummarizer? = tags.summarizer(tag)

    override fun nameForValue(value: ULong): String = tags.nameForValue(value)
}

/**
 * Thread-safe global format context singleton.
 *
 * Lazily initialized with standard tags, known values, functions, and parameters.
 */
object GlobalFormatContext {
    private val context: FormatContext by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        // Ensure component tags are registered in the global dcbor tag store
        com.blockchaincommons.bccomponents.registerTags()

        // Build our own TagsStore for the FormatContext
        val tags = TagsStore()

        val ctx = FormatContext(
            tags = tags,
            knownValues = KNOWN_VALUES,
            functions = GLOBAL_FUNCTIONS,
            parameters = GLOBAL_PARAMETERS,
        )
        registerTagsIn(ctx)
        ctx
    }

    /** Returns the global format context. */
    fun get(): FormatContext = context
}

/**
 * Accesses the global format context for read-only operations.
 */
fun <T> withFormatContext(action: (FormatContext) -> T): T {
    return action(GlobalFormatContext.get())
}

/**
 * Registers standard tags and summarizers in a format context.
 */
fun registerTagsIn(context: FormatContext) {
    // Register standard component tags into the context's tag store
    com.blockchaincommons.bccomponents.registerTagsIn(context.tags())

    // Known value summarizer
    val knownValues = context.knownValues()
    context.tags().setSummarizer(TAG_KNOWN_VALUE) { untaggedCbor, _ ->
        val kv = KnownValue.fromUntaggedCbor(untaggedCbor)
        knownValues.name(kv).flankedBy("'", "'")
    }

    // Function summarizer
    val functions = context.functions()
    context.tags().setSummarizer(TAG_FUNCTION) { untaggedCbor, _ ->
        val f = Function.fromUntaggedCbor(untaggedCbor)
        FunctionsStore.nameForFunction(f, functions).flankedBy("\u00AB", "\u00BB")
    }

    // Parameter summarizer
    val parameters = context.parameters()
    context.tags().setSummarizer(TAG_PARAMETER) { untaggedCbor, _ ->
        val p = Parameter.fromUntaggedCbor(untaggedCbor)
        ParametersStore.nameForParameter(p, parameters).flankedBy("\u2770", "\u2771")
    }

    // Request summarizer
    context.tags().setSummarizer(TAG_REQUEST) { untaggedCbor, flat ->
        val e = Envelope.newLeaf(untaggedCbor)
        val formatted = e.formatOpt(flat = flat, context = FormatContextOpt.Custom(context))
        formatted.flankedBy("request(", ")")
    }

    // Response summarizer
    context.tags().setSummarizer(TAG_RESPONSE) { untaggedCbor, flat ->
        val e = Envelope.newLeaf(untaggedCbor)
        val formatted = e.formatOpt(flat = flat, context = FormatContextOpt.Custom(context))
        formatted.flankedBy("response(", ")")
    }

    // Event summarizer
    context.tags().setSummarizer(TAG_EVENT) { untaggedCbor, flat ->
        val e = Envelope.newLeaf(untaggedCbor)
        val formatted = e.formatOpt(flat = flat, context = FormatContextOpt.Custom(context))
        formatted.flankedBy("event(", ")")
    }
}

/**
 * Registers standard tags in the global format context.
 */
fun registerTags() {
    // Accessing the global format context triggers lazy initialization
    // which calls registerTagsIn
    GlobalFormatContext.get()
}

File: csharp/BCEnvelope/BCEnvelope/FormatContext.cs
using BlockchainCommons.BCComponents;
using BlockchainCommons.BCTags;
using BlockchainCommons.DCbor;
using BlockchainCommons.KnownValues;

namespace BlockchainCommons.BCEnvelope;

/// <summary>
/// Context object for formatting Gordian Envelopes with annotations.
/// </summary>
/// <remarks>
/// Provides information about CBOR tags, known values, functions, and parameters
/// used to annotate the output of envelope formatting functions.
/// </remarks>
public sealed class FormatContext
{
    private readonly TagsStore _tags;
    private readonly KnownValuesStore _knownValues;
    private readonly FunctionsStore _functions;
    private readonly ParametersStore _parameters;

    /// <summary>
    /// Creates a new format context with the specified components.
    /// </summary>
    public FormatContext(
        TagsStore? tags = null,
        KnownValuesStore? knownValues = null,
        FunctionsStore? functions = null,
        ParametersStore? parameters = null)
    {
        _tags = tags ?? new TagsStore();
        _knownValues = knownValues ?? new KnownValuesStore();
        _functions = functions ?? new FunctionsStore();
        _parameters = parameters ?? new ParametersStore();
    }

    /// <summary>Returns the CBOR tags registry.</summary>
    public TagsStore Tags => _tags;

    /// <summary>Returns the known values registry.</summary>
    public KnownValuesStore KnownValues => _knownValues;

    /// <summary>Returns the functions registry.</summary>
    public FunctionsStore Functions => _functions;

    /// <summary>Returns the parameters registry.</summary>
    public ParametersStore Parameters => _parameters;

    // --- TagsStore delegation ---

    /// <summary>Returns the assigned name for a tag if one exists.</summary>
    public string? AssignedNameForTag(Tag tag) => _tags.AssignedNameForTag(tag);

    /// <summary>Returns a name for a tag, either the assigned name or a generic representation.</summary>
    public string NameForTag(Tag tag) => _tags.NameForTag(tag);

    /// <summary>Looks up a tag by its name.</summary>
    public Tag? TagForName(string name) => _tags.TagForName(name);

    /// <summary>Looks up a tag by its numeric value.</summary>
    public Tag? TagForValue(ulong value) => _tags.TagForValue(value);

    /// <summary>Returns a CBOR summarizer for a tag value if one exists.</summary>
    public CborSummarizer? GetSummarizer(ulong tagValue) => _tags.GetSummarizer(tagValue);

    /// <summary>Returns a name for a tag value.</summary>
    public string NameForValue(ulong value) => _tags.NameForValue(value);
}

/// <summary>
/// Thread-safe global format context singleton.
/// </summary>
public static class GlobalFormatContext
{
    private static readonly object _lock = new();
    private static FormatContext? _context;

    /// <summary>
    /// Gets the global format context, initializing it if necessary.
    /// </summary>
    public static FormatContext Get()
    {
        if (_context is not null)
            return _context;

        lock (_lock)
        {
            if (_context is not null)
                return _context;

            // Ensure component tags are registered in the global dcbor tag store
            TagsRegistry.RegisterTags();

            var ctx = new FormatContext(
                tags: new TagsStore(),
                knownValues: KnownValuesRegistry.KnownValues.Get(),
                functions: Functions.GlobalStore,
                parameters: Parameters.GlobalStore);

            RegisterTagsIn(ctx);
            _context = ctx;
            return _context;
        }
    }

    /// <summary>
    /// Executes an action with read access to the global format context.
    /// </summary>
    public static T WithFormatContext<T>(Func<FormatContext, T> action)
    {
        return action(Get());
    }

    /// <summary>
    /// Executes an action with the global format context (void version).
    /// </summary>
    public static void WithFormatContext(Action<FormatContext> action)
    {
        action(Get());
    }

    /// <summary>
    /// Registers standard tags and summarizers in a format context.
    /// </summary>
    public static void RegisterTagsIn(FormatContext context)
    {
        // Register standard component tags
        TagsRegistry.RegisterTagsIn(context.Tags);

        // Known value summarizer
        var knownValues = context.KnownValues;
        context.Tags.SetSummarizer(BcTags.TagKnownValue, (untaggedCbor, _) =>
        {
            var kv = KnownValue.FromUntaggedCbor(untaggedCbor);
            return knownValues.Name(kv).FlankedBy("'", "'");
        });

        // Function summarizer
        var functions = context.Functions;
        context.Tags.SetSummarizer(BcTags.TagFunction, (untaggedCbor, _) =>
        {
            var f = Function.FromUntaggedCbor(untaggedCbor);
            return FunctionsStore.NameForFunction(f, functions).FlankedBy("\u00AB", "\u00BB");
        });

        // Parameter summarizer
        var parameters = context.Parameters;
        context.Tags.SetSummarizer(BcTags.TagParameter, (untaggedCbor, _) =>
        {
            var p = Parameter.FromUntaggedCbor(untaggedCbor);
            return ParametersStore.NameForParameter(p, parameters).FlankedBy("\u2770", "\u2771");
        });

        // Request summarizer
        var clonedContext1 = context;
        context.Tags.SetSummarizer(BcTags.TagRequest, (untaggedCbor, flat) =>
        {
            var e = Envelope.CreateLeaf(untaggedCbor);
            var formatted = e.FormatOpt(new EnvelopeFormatOpts(flat, FormatContextOpt.Custom(clonedContext1)));
            return formatted.FlankedBy("request(", ")");
        });

        // Response summarizer
        var clonedContext2 = context;
        context.Tags.SetSummarizer(BcTags.TagResponse, (untaggedCbor, flat) =>
        {
            var e = Envelope.CreateLeaf(untaggedCbor);
            var formatted = e.FormatOpt(new EnvelopeFormatOpts(flat, FormatContextOpt.Custom(clonedContext2)));
            return formatted.FlankedBy("response(", ")");
        });

        // Event summarizer
        var clonedContext3 = context;
        context.Tags.SetSummarizer(BcTags.TagEvent, (untaggedCbor, flat) =>
        {
            var e = Envelope.CreateLeaf(untaggedCbor);
            var formatted = e.FormatOpt(new EnvelopeFormatOpts(flat, FormatContextOpt.Custom(clonedContext3)));
            return formatted.FlankedBy("event(", ")");
        });
    }

    /// <summary>
    /// Registers standard tags in the global format context.
    /// </summary>
    public static void RegisterTags()
    {
        // Accessing the global format context triggers initialization
        Get();
    }
}
